src/main/java/com/exasol/adapter/document/documentfetcher/files/randomaccessinputstream/RandomAccessInputStreamCache.java

```diff
@@ -0,0 +1,111 @@
+        this.position = position;
+    }
+
+    @Override
+    public long getPos() {
+        return this.position;
+    }
+
+    @Override
+    public long getLength() throws IOException {
+        return this.source.getLength();
+    }
+
+    @Override
+    public int read() throws IOException {
+        if (this.position < this.getLength()) {
+            if (!isCurrentPositionInCache()) {
+                fillCache();
+            }
+            final long positionInCache = this.position - this.cacheStartPos;
+            this.position++;
+            return this.cache[(int) positionInCache] & 0xFF;
+        } else {
+            return -1;
+        }
+    }
+
+    private void fillCache() throws IOException {
+        this.source.seek(this.position);
+        this.usedCacheSize = this.source.read(this.cache, 0, this.cache.length);
+        this.cacheStartPos = this.position;
+    }
+
+    @Override
+    public int read(final byte[] destination, final int offset, final int length) throws IOException {
+        if (length == 0) {
+            return 0;
+        }
+        final long remainingBytesInStream = getLength() - this.position;
+        final int actualReadLength = (int) Math.min(length, remainingBytesInStream);
+        if (actualReadLength > 0) {
+            final long positionInCache = this.position - this.cacheStartPos;
+            final long cachedSize = getCachedSize(actualReadLength, positionInCache);
+            if (positionInCache >= 0 && cachedSize > 0) {
+                System.arraycopy(this.cache, (int) positionInCache, destination, offset, (int) cachedSize);
+                this.position += cachedSize;
+            }
+            final long remainingSize = actualReadLength - cachedSize;
+            if (remainingSize > 0) {
+                return (int) cachedSize + readRemaining(destination, offset, cachedSize, remainingSize);
+            }
+        }
+        return actualReadLength == 0 ? -1 : actualReadLength;
+    }
+
+    private long getCachedSize(final int actualReadLength, final long positionInCache) {
+        if (positionInCache < 0) {
+            return 0;
+        } else {
+            return Math.max(Math.min(actualReadLength, this.usedCacheSize - positionInCache), 0);
+        }
+    }
+
+    private int readRemaining(final byte[] destination, final int offset, final long cachedSize,
+            final long remainingSize) throws IOException {
+        this.source.seek(this.position);
+        if (remainingSize < this.cache.length) {
+            final int bytesRead = this.source.read(this.cache, 0, this.cache.length);
+            final long requestedReadBytes = Math.min(bytesRead, remainingSize);
+            System.arraycopy(this.cache, 0, destination, (int) (offset + cachedSize), (int) requestedReadBytes);
+            this.cacheStartPos = this.position;
+            this.position += requestedReadBytes;
+            this.usedCacheSize = bytesRead;
+            return (int) requestedReadBytes;
```

Comment:
I want to understand what happens in the first part of the if-else statement.
It fills the cache, but there maybe some left overs in the cache array. But this does not happen. 
That is, if we still have remainings fewer than cache size, we can fill the cache, copy required and update positions. Otherwise, if remainings larger than cache, directly read to destination.